File: src/JVSE.c
#include "JVSE.h"

void intHandler(int dummy) {
  closeKeyboard();
  close(serial);
  exit(0);
}

int main(void) {
    /* Setup signal handlers */
    signal(SIGINT, intHandler);

    /* Print out information */
    printf("JVSE: OpenJVS Emulator %d.%d\n", majorVersion, minorVersion);
    printf("JVSE: (C) Robert Dilley 2018\n\n");

    /* Setup the serial interface here */
    serial = open(portName, O_RDWR | O_NOCTTY | O_SYNC | O_NONBLOCK);
    
    if (serial < 0) {
        printf("JVSE: Failed to connect\n");
        return -1;
    }
    set_interface_attribs(serial, B115200);

    /* Init the modules here */
    initControl();
    initConfig();

    /* Setup some example keys */
    setSystemSwitch(5, 0);
    setSystemSwitch(4, 0);
    setAnalogue(0, 0xFF);
    setPlayerSwitch(0, 1, 0);
    setPlayerSwitch(0, 9, 0);

    /* Run the system forever */
    while (1) {
        getPacket();
    }

    return 0;
}

void debug(char* string) {
	//printf("%s", string);
}

/* Write the byte to the serial buffer adding appropriate escape bytes */
void writeEscaped(unsigned char byte) {
    if (byte == CMD_SYNC || byte == CMD_ESCAPE) {
        unsigned char buffer[] = {
            CMD_ESCAPE
        };
        int n = write(serial, buffer, sizeof(buffer));
        if (n != 1) {
            printf("JVSE: Error from write: %d, %d\n", n, errno);
        }
        byte -= 1;
    }

    unsigned char buffer[] = {
        byte
    };
    int n = write(serial, buffer, sizeof(buffer));
    if (n != 1) {
        printf("JVSE: Error from write: %d, %d\n", n, errno);
    }
    //usleep(10);
    //tcdrain(serial);
}

unsigned char getByte() {
    unsigned char buffer[] = {
        0x00
    };
    int n = -1;
    while (n < 1) {
        //usleep(10);
        n = read(serial, buffer, 1);
    }

    if (buffer[0] == CMD_ESCAPE) {
        n = -1;
        while (n < 1) {
            //usleep(10);
            n = read(serial, buffer, 1);
        }
        return buffer[0] + 1;
    }
    return buffer[0];
}

void writeByte(unsigned char byte) {
    reply[replyCount] = byte;
    replyCount++;
}

void writeString(char * string) {
    for (int i = 0; i < strlen(string); i++) {
        reply[replyCount] = string[i];
        replyCount++;
    }
}

void writeBytes(unsigned char bytes[], int size) {
    for (int i = 0; i < size; i++) {
        reply[replyCount] = bytes[i];
        replyCount++;
    }
}

void sendReply() {
    //usleep(100);
    if (replyCount > 0) {
        int checksum = BUS_MASTER + replyCount + 2 + STATUS_SUCCESS;

				/* Write out the sync packet */
        unsigned char buffer[] = {
            CMD_SYNC
        };
        write(serial, buffer, 1);

        writeEscaped(BUS_MASTER);
        writeEscaped(replyCount + 2);
        writeEscaped(STATUS_SUCCESS);

        int count = 0;
        while (count < replyCount) {
            checksum += reply[count];
            writeEscaped(reply[count]);
            count++;
        }

        writeEscaped(checksum & 0xFF);

        replyCount = 0;
    }

    //tcflush(serial, TCIOFLUSH);
    //tcdrain(serial);
}

void processPacket(unsigned char packet[], int packet_length, int packet_address) {
    if (packet_address == CMD_BROADCAST || packet_address == deviceID || allDeviceMode) {
        while (packet_length > 0) {
            int command_size = 1;

            if (packet[0] == CMD_RESET) {
                debug("CMD_RESET\n");
                command_size = 2;
                deviceID = -1;
            } else if (packet[0] == CMD_SETADDRESS) {
                debug("CMD_SETADDRESS\n");
                command_size = 2;
                deviceID = packet[1];
                writeByte(STATUS_SUCCESS);
            } else if (packet[0] == CMD_READID) {
                debug("CMD_READID\n");
                writeByte(STATUS_SUCCESS);
                writeString("JVSE Emulator;I/O BD JVS;837-13551;Ver1.00;98/10");
            } else if (packet[0] == CMD_FORMATVERSION) {
                debug("CMD_FORMATVERSION\n");
                writeByte(STATUS_SUCCESS);
                writeByte(0x11);
            } else if (packet[0] == CMD_JVSVERSION) {
                debug("CMD_JVSVERSION\n");
                writeByte(STATUS_SUCCESS);
                writeByte(0x20);
            } else if (packet[0] == CMD_COMMSVERSION) {
                debug("CMD_COMMSVERSION\n");
                writeByte(STATUS_SUCCESS);
                writeByte(0x10);
            } else if (packet[0] == CMD_GETFEATURES) {
                debug("CMD_GETFEATURES\n");
                writeByte(STATUS_SUCCESS);
                unsigned char features[] = {
                    0x01, players, bytesPerPlayer * 8, 0x00,
                    0x02, 0x02, 0x00, 0x00,
                    0x03, analogueChannels, 0x08, 0x00,
                    0x06, 0x08, 0x08, 0x02,
                    0x00
                };
                writeBytes(features, sizeof(features));
            } else if (packet[0] == CMD_READSWITCHES) {
                debug("CMD_READSWITCHES\n");
                writeByte(STATUS_SUCCESS);
                writeByte(reverse(systemSwitches));
								if(packet[1] != players || packet[2] != bytesPerPlayer) {
									printf("JVSE: Switch request differs from offered no. of players\n");
								}
								for(int i = 0 ; i < packet[1] * packet[2] ; i++) {
										writeByte(reverse(playerSwitches[i]));
								}
                command_size = 3;
            } else if (packet[0] == CMD_WRITEGPIO1) {
                debug("CMD_WRITEGPIO1\n");
                command_size = 2 + packet[1];
                writeByte(STATUS_SUCCESS);
            } else if (packet[0] == CMD_SETMAINBOARDID) {
                debug("CMD_MAINBOARDID\n");
                int counter = 1;
                while (packet[counter] != 0x00 && counter <= packet_length) {
										boardID[counter - 1] = packet[counter];
                    counter++;
                }
								boardID[counter - 1] = 0x00;
                command_size = counter;
                writeByte(STATUS_SUCCESS);
            } else if (packet[0] == CMD_READCOIN) {
                debug("CMD_READCOIN\n");
                writeByte(STATUS_SUCCESS);
                writeByte(0x00);
                writeByte(coin);
                writeByte(0x00);
                writeByte(0x00);
                command_size = 2;
            } else if (packet[0] == CMD_READANALOG) {
                debug("CMD_READANALOG\n");
                command_size = 2;
                writeByte(STATUS_SUCCESS);

								if(packet[1] != analogueChannels) {
									printf("JVSE: Analogue Channel Requests differs\n");
								}

								for(int i = 0 ; i < packet[1] ; i++) {
									writeByte(0x00);
									writeByte(analogue[i]);
								}
            } else if (packet[0] == CMD_READSCREENPOS) {
                debug("CMD_READSCREENPOS\n");
                command_size = 2;
                writeByte(STATUS_SUCCESS);
                writeByte(0x00);
                writeByte(analogue[0]);
                writeByte(0x00);
                writeByte(analogue[1]);
						} else if (packet[0] == CMD_WRITECOINSUBTRACT) {
							command_size = 4;
							writeByte(STATUS_SUCCESS);
							coin -= packet[3];
            } else {
                printf("CMD_UNKNOWN %02hhX \n", packet[0]);
            }

            packet_length -= command_size;
            packet += command_size;
        }
        //usleep(100);
        sendReply();
    }
}

void getPacket() {
    int ourChecksum = 0;

    while (getByte() != CMD_SYNC);
    //printf("SYNC\n");
    unsigned char packet_address = getByte();
    ourChecksum += packet_address;

    unsigned char packet_length = getByte() - 1;
    ourChecksum += packet_length + 1;
    //printf("length %d\n", packet_length);
    unsigned char packet[packet_length];

    int counter = 0;
    while (counter < packet_length) {
        unsigned char byte = getByte();
        packet[counter] = byte;
        //printf("byte %d\n", byte);
        counter += 1;
        ourChecksum += byte;
    }

    unsigned char theirChecksum = getByte();
    //printf("check %d %d\n", theirChecksum, (ourChecksum & 0xFF));
    if ((ourChecksum & 0xFF) == theirChecksum) {
        processPacket(packet, packet_length, packet_address);
    } else {
        printf("JVSE: Error Checksum Problem\n");
    }

}

File: OpenJVS-0.1/Python/JVSE/JVSE.py
import serial
import threading

# Set the USB Address here
USB_UART_ADDRESS = '/dev/tty.SLAB_USBtoUART'

# List of all the JVS standard commands
CMD_BROADCAST = 0xFF
CMD_ESCAPE = 0xD0
CMD_SYNC = 0xE0
CMD_RESET = 0xF0
CMD_SETADDRESS = 0xF1
CMD_SETMETHOD = 0xF2
CMD_READID = 0x10
CMD_FORMATVERSION = 0x11
CMD_JVSVERSION = 0x12
CMD_COMMSVERSION = 0x13
CMD_GETFEATURES = 0x14
CMD_SETMAINBOARDID = 0x15
CMD_READSWITCHES = 0x20
CMD_READCOIN = 0x21
CMD_READANALOG = 0x22
CMD_READROTARY = 0x23
CMD_READKEYCODE = 0x24
CMD_READSCREENPOS = 0x25
CMD_READGPIO = 0x26
CMD_WRITEPAYOUTREMAINING = 0x2E
CMD_RESEND = 0x2F
CMD_WRITECOINSUBTRACT = 0x30
CMD_WRITEPAYOUT = 0x31
CMD_WRITEGPIO1 = 0x32
CMD_WRITEANALOG = 0x33
CMD_WRITECHAR = 0x34
CMD_WRITECOINADDED = 0x35
CMD_WRITEPAYOUTSUBTRACT = 0x36
CMD_WRITEGPIOBYTE = 0x37
CMD_WRITEGPIOBIT = 0x38
BUS_MASTER	= 0x00
STATUS_SUCCESS = 0x01

class JVSE(threading.Thread):
    """A class to provide control of an arcade board via the JVS standard.
    Requires a USB to RS485 converter.

    Attributes:
         serial_path: The path to the RS485 converter device.
    """

    def __init__(self, serial_path = USB_UART_ADDRESS):
        """Return a JVS object relaying JVS commands to the device at *serial_path*."""
        threading.Thread.__init__(self)
        self.serial = serial.Serial(serial_path, 115200, timeout=5)
        self.old_reply = []
        self.device_id = 1
        self.running = False
        self.system_key = [0] * 8
        self.player1_key = [0] * 16
        self.player2_key = [0] * 16
        self.coin = 0
        self.posx = 0x00
        self.posy = 0x00

    def increment_coin(self):
        """Increment the amount of coins reported by the first coin slot."""
        self.coin += 1

    def set_key_player1(self, number, value):
        """Set player 1's switchs defined by *number* to either pressed (1) or unpressed (0) using *value*."""
        self.player1_key[number] = value

    def set_key_player2(self, number, value):
        """Set player 2's switchs defined by *number* to either pressed (1) or unpressed (0) using *value*."""
        self.player2_key[number] = value

    def set_key_system(self, number, value):
        """Set the systems switchs defined by *number* to either pressed (1) or unpressed (0) using *value*."""
        self.system_key[number] = value

    def set_pos(self, x, y):
        """Set player 1's lightgun and analogue position defined by *x* and *y*."""
        self.posx = x
        self.posy = y

    def run(self):
        """Start the JVS connection thread."""
        self.running = True
        while self.running:
            self.get_packet()

    def stop(self):
        """Stop the JVS connection thread."""
        self.running = False

    def bits_to_byte(self, bit_list):
        """Convert a list of bits *bit_list*, to a single byte."""
        byte = 0x0
        mult = 1
        for i in range(len(bit_list) -1, -1, -1):
            if bit_list[i] == 1:
                byte += mult
            mult *= 2
        return byte

    def write_escaped(self, byte):
        """Write the byte *byte* to the serial buffer adding appropriate escape bytes."""
        if byte == CMD_SYNC or byte == CMD_ESCAPE:
            self.serial.write(chr(CMD_ESCAPE))
            byte -= 1
        print "Write: ", byte
        self.serial.write(chr(byte))

    def write_reply(self, reply):
        """Write the byte list *reply* to the serial buffer."""
        if len(reply) > 0:
            checksum = BUS_MASTER + (len(reply) + 2) + STATUS_SUCCESS
            self.serial.write(chr(CMD_SYNC))
            self.write_escaped(BUS_MASTER)
            self.write_escaped(len(reply) + 2)
            self.write_escaped(STATUS_SUCCESS)
            for i in range(0, len(reply)):
                checksum += reply[i]
                self.write_escaped(reply[i])
            self.write_escaped(checksum & 0xFF)
            old_reply = reply
            reply = []
            self.serial.flush()


    def get_byte(self):
        """Get a single byte from the serial buffer, removing any escape bytes."""
        byte = self.serial.read(1)
        if len(byte) == 0:
            return self.get_byte()
        byte = ord(byte)
        if byte == CMD_ESCAPE:
            byte = self.serial.read(1)
            if len(byte) == 0:
                # Maybe should be plus 1?
                return self.get_byte() + 1
            else:
                return ord(byte) + 1
        return byte

    def process_packet(self, packet, packet_address):
        """Process the packet *packet* sent to *packet_address* and send the appropriate reply"""
        print "Process Packet: ", packet
        if packet_address == CMD_BROADCAST or packet_address == self.device_id:
            packet_length = len(packet)
            reply = []
            while packet_length > 0:
                command_size = 1

                if packet[0] == CMD_RESET:
                    print "CMD_RESET"
                    command_size = 2
                    self.device_id = -1

                elif packet[0] == CMD_SETADDRESS:
                    print "CMD_SETADDRESS"
                    command_size = 2
                    self.device_id = packet[1]
                    reply += [STATUS_SUCCESS]

                elif packet[0] == CMD_READID:
                    print "CMD_READID"
                    reply += [STATUS_SUCCESS] + map(ord, "SEGA ENTERPRISESLTD.;I/O BD JVS;837-13551;Ver1.00;98/10") + [0x00]

                elif packet[0] == CMD_FORMATVERSION:
                    print "CMD_FORMATVERSION"
                    reply += [STATUS_SUCCESS, 0x11]

                elif packet[0] == CMD_JVSVERSION:
                    print "CMD_JVSVERSION"
                    reply += [STATUS_SUCCESS, 0x20]

                elif packet[0] == CMD_COMMSVERSION:
                    print "CMD_COMMAVERSION"
                    reply += [STATUS_SUCCESS, 0x10]

                elif packet[0] == CMD_GETFEATURES:
                    print "CMD_GETFEATURES"
                    features = [
                                0x01, 0x02, 0x0B, 0x00,
                                0x02, 0x02, 0x00, 0x00,
                                0x03, 0x04, 0x08, 0x00,
                                0x06, 0x08, 0x08, 0x02,
                                0x00
                                ]
                    reply += [STATUS_SUCCESS] + features

                elif packet[0] == CMD_WRITEGPIO1:
                    command_size = 2 + packet[1]
                    reply += [STATUS_SUCCESS]

                elif packet[0] == CMD_SETMAINBOARDID:
                    is_not_blank = 1
                    counter = 1
                    while is_not_blank:
                        if ord(packet[counter]) == 0x00 or ord(packet[counter]) == 0 or counter >= packet_length:
                            is_not_blank = 0
                        else:
                            counter+=1
                    command_size = counter
                    reply += [STATUS_SUCCESS]

                elif packet[0] == CMD_READCOIN:
                    reply += [STATUS_SUCCESS] + [0x00] + [self.coin] + [0x00] + [0x00]
                    command_size = 2

                elif packet[0] == CMD_READSWITCHES:
                    sw = [STATUS_SUCCESS] + [self.bits_to_byte(self.system_key)] + [self.bits_to_byte(self.player1_key[0:8])] + [self.bits_to_byte(self.player1_key[8:16])] + [self.bits_to_byte(self.player2_key[0:8])] + [self.bits_to_byte(self.player2_key[8:16])]
                    reply += sw
                    command_size = 3

                elif packet[0] == CMD_READANALOG:
                    command_size = 2
                    reply += [STATUS_SUCCESS] + [self.posx & 0xFF] + [0x00] + [self.posy & 0xFF] + [0x00] + [self.posx & 0xFF] + [0x00] + [self.posy & 0xFF] + [0x00]

                elif packet[0] == CMD_READSCREENPOS:
                    command_size = 2
                    reply += [STATUS_SUCCESS] + [0x00] + [self.posx & 0xFF] + [0x00] + [self.posy & 0xFF]

                else:
                    print "Error: Packet Unknown: ", hex(packet[0])

                packet_length -= command_size
                packet = packet[command_size:]

            self.write_reply(reply)

    def get_packet(self):
        """Get and process a packet"""
        while self.get_byte() != CMD_SYNC:
            pass

        checksum = 0
        packet = []

        packet_address = self.get_byte()
        checksum += packet_address
        print "Address: ", packet_address

        packet_length = self.get_byte() - 1
        checksum += packet_length + 1
        print "Length: ", packet_length

        while len(packet) < packet_length:
            byte = self.get_byte()
            checksum += byte
            packet += [byte]

        if (checksum & 0xFF) == self.get_byte():
            self.process_packet(packet, packet_address)
        else:
            print "Error: Checksum Problem"
